test/e2e/framework-tests/framework-tests.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package frameworktests
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+)
+
+// timeout must be large enough that we can distinguish it from an immediate
+// return of the poll function.
+var timeout = 20 * time.Second
+
+func SIGDescribe(text string, body func()) bool {
+	return ginkgo.Describe("[sig-testing] "+text, body)
+}
+
+var _ = SIGDescribe("E2E framework", func() {
```

Comment:
I'll try that approach.

Reply:
Hmm, are those unit tests supposed to pass?

```
$ go test ./test/e2e/framework
Running Suite: Logging Suite
============================
Random Seed: 1652257205
Will run 4 of 4 specs

• Failure [0.002 seconds]
log
/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log_test.go:50
  fails [It]
  /nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log_test.go:54

  May 11 10:20:05.325: I'm failing.

  /nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log_test.go:57
...
Ran 4 of 4 Specs in 0.010 seconds
FAIL! -- 0 Passed | 4 Failed | 0 Pending | 0 Skipped
--- FAIL: TestFailureOutput (0.02s)
panic: 
Your test failed.
Ginkgo panics to prevent subsequent assertions from running.
Normally Ginkgo rescues this panic so you shouldn't see it.

But, if you make an assertion in a goroutine, Ginkgo can't capture the panic.
To circumvent this, you should call

	defer GinkgoRecover()

at the top of the goroutine that caused this panic.
 [recovered]
	panic: 
Your test failed.
Ginkgo panics to prevent subsequent assertions from running.
Normally Ginkgo rescues this panic so you shouldn't see it.

But, if you make an assertion in a goroutine, Ginkgo can't capture the panic.
To circumvent this, you should call

	defer GinkgoRecover()

at the top of the goroutine that caused this panic.
 [recovered]
	panic: 
Your test failed.
Ginkgo panics to prevent subsequent assertions from running.
Normally Ginkgo rescues this panic so you shouldn't see it.

But, if you make an assertion in a goroutine, Ginkgo can't capture the panic.
To circumvent this, you should call

	defer GinkgoRecover()

at the top of the goroutine that caused this panic.


goroutine 130 [running]:
testing.tRunner.func1.2({0x19cca20, 0xc000915680})
	/nvme/gopath/go-1.18.1/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/nvme/gopath/go-1.18.1/src/testing/testing.go:1392 +0x39f
panic({0x19cca20, 0xc000915680})
	/nvme/gopath/go-1.18.1/src/runtime/panic.go:838 +0x207
k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.Fail.func1()
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper/wrapper.go:63 +0x7d
panic({0x17ca9e0, 0x1d8a140})
	/nvme/gopath/go-1.18.1/src/runtime/panic.go:838 +0x207
github.com/onsi/ginkgo.Fail({0xc0007c26c0, 0x8e}, {0xc0000298f0?, 0xc000029900?, 0x0?})
	/nvme/gopath/pkg/mod/github.com/onsi/ginkgo@v1.14.0/ginkgo_dsl.go:260 +0xdd
k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.Fail({0xc0007c26c0, 0x8e}, {0xc0000299d0?, 0x1afce88?, 0xc0000299f0?})
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper/wrapper.go:67 +0x197
k8s.io/kubernetes/test/e2e/framework.Fail({0xc0008ca700, 0x79}, {0xc0008b9a30?, 0xc00091da40?, 0xc000029a90?})
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log.go:63 +0x145
github.com/onsi/gomega/internal/assertion.(*Assertion).match(0xc000029b40, {0x1d99e20, 0xc0008e4310}, 0x1, {0xc000029cb8, 0x2, 0x2})
	/nvme/gopath/pkg/mod/github.com/onsi/gomega@v1.10.1/internal/assertion/assertion.go:79 +0x1bd
github.com/onsi/gomega/internal/assertion.(*Assertion).To(0xc000029b40, {0x1d99e20, 0xc0008e4310}, {0xc000029cb8, 0x2, 0x2})
	/nvme/gopath/pkg/mod/github.com/onsi/gomega@v1.10.1/internal/assertion/assertion.go:38 +0x95
k8s.io/kubernetes/test/e2e/framework.ExpectEqual({0x17ca9e0?, 0xc0008e42f0?}, {0x17ca9e0?, 0xc0008e4300}, {0xc000029cb8, 0x2, 0x2})
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/expect.go:29 +0x125
k8s.io/kubernetes/test/e2e/framework_test.TestFailureOutput(0x0?)
	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/log_test.go:123 +0x77b
testing.tRunner(0xc000600d00, 0x1c0d208)
	/nvme/gopath/go-1.18.1/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/nvme/gopath/go-1.18.1/src/testing/testing.go:1486 +0x35f
FAIL	k8s.io/kubernetes/test/e2e/framework	0.070s
FAIL
```

Those tests are not run by "make test". Not sure whether that is intentional?

https://github.com/kubernetes/kubernetes/blob/b74d023e70d6064c7f3f77031e7d26ec38497fc9/hack/make-rules/test.sh#L47

Reply:
I can't see the goroutine that panics :thinking:

Reply:
It's the goroutine which executes TestFailureOutput. "go test" internally uses multiple different goroutines.

Reply:
Fix is in https://github.com/kubernetes/kubernetes/pull/109961

Reply:
> I'll try that approach.

It works. PR #109961 must be merged first, then I can follow with tests for the current e2epod code. Here's a preview: https://github.com/kubernetes/kubernetes/compare/master...pohly:pod-wait-tests?expand=1

@tallclair: can you help get that PR merged?

Reply:
I just realized something... I'm the one who wrote this `log_test.go`. I had completely forgotten about that. :sweat_smile: